### PR TITLE
TodoMVC example refactoring using `route` component

### PR DESCRIPTION
**Pull request by yosbelms**
The usage of the `route` component made cleaner the `Footer` view, removing some logic of `onclick` events and delegating the control to `href` + router. The state holder (`currentFilter`) has been moved to the model (where it should have been from the beginning). Althoug the conditional in the router initializer that checks if no route is provided may looks a bit clunky because `domvm.route` doesn't offers API to do that checking AFAIK.

### CODE REVIEW

**Comment by leeoniya** [comment]
thanks @yosbelms :)

Ideally, the view/app could utilize the router to generate the hrefs (which could be parameterized as needed)

```js
['a', {href: router.href("all"), class: makeCls('all')}, 'All']
```

This touches on a bigger discussion regarding concern separation and SPA structuring in domvm. I'd like to put some thoughts here and see what you think.

When I write apps, I prefer to provide full control via API. This means that the address bar and view have the dual task of both reflecting internal state (which can change with manual API method invocation) as well as setting it via input interaction. For example, following a nav link to `#/orders` should invoke `app.setModule("orders")`. Likewise, manually invoking `app.setModule("orders")` should set the url/history to `#/orders`. This bi-directional operation allows for full automation via self-contained method calls without always remembering to string together model state activation, view switching, and route setting. In addition, the views should be able to use the router to generate hrefs. 

Full decoupling also causes a chicken-egg problem (should the router be created first or the app/view since they use each other for various things). With this in mind, my preference is to have apps somewhat coupled to allow the the model, view and router to freely interact in a global-esque fashion. More concretely, (probably non-functional but pseudo):


```js
function App(state, opts) {
	this.state = state;
	
	// router and view can interact via the app
	this.router = domvm.route(AppRouter, this);
	this.view = domvm.view(AppView, this);
	
	this.curModule = null;
}

App.prototype = {
	setModule: function(name) {
		// ajax/fetch state, whatever
		this.curModule = new OrdersModule();
		// set url
		router.goto(name);
	}
};

function AppView(vm, app) {
	return function() {
		return ["body#app",
			["header",
				["nav#tabs",
					["ul",
						["li",
							["a", {href: app.router.href("orders")}, "Orders"]
						]
					]
				]
			],
			["main",
				this.curModule.view
			],
			["footer", "Footer..."],
		];
	}
}

function AppRouter(router, app) {
	return {
		orders: {
			path: "/orders",
			onenter: function() {
				app.setModule("orders");
			}
		}
	};
}

funtion OrdersModule(state, app) {
	this.app = app;
	this.view = domvm.view(OrdersModuleView, this);
}

function OrdersModuleView(vm, orders) {
	return function() {
		return ["section.module",
			["h1", "Orders Module"]
		];
	};
}

var app = new App();

app.router.refresh();
```

> Although the conditional in the router initializer that checks if no route is provided may looks a bit clunky because domvm.route doesn't offers API to do that checking AFAIK.

getting sleepy, but i'll get back to you on this :)

**Comment by yosbelms** [comment]
Apps using domvm could be crafted using the architecture you mentioned above, but architectures gradually evolves and often is a determined in teams, or by a CTO. That said, if domvm provides tools to organize applications, fine, but should be decoupled enough to adapt to the needs of individuals.

The TodoMVC is just a example to illustrate domvm usage as simple as possible. Let's think as if we are in the shoes of the new domvm users, for example, due to the simplicity of the routing logic I could write:

```js
var ret = {};

['all', 'active', 'completed'].forEach(function(r) {
    ret[r] = {
        path: r,
        onenter: (function(r) {
            return function() {
                model.Todos.currentFilter = r;
                appView.redraw();
            }
        })(r)
    }
});

return ret;
```

But with the above code is hard to figure out what is the domvm API for routing.

**Comment by leeoniya** [comment]
My main concern is with positioning the router as the gatekeeper & glue to app actions such as item filtering. The design forces interaction with the app via the address bar, links or via route dispatching to get view redraws. It means you cannot take this TodoList app and turn it into an in-page widget that can switch views without relying on urls or polluting browsing history.

By keeping as much as possible contained and linked within the model, your router mearly becomes a decorator for the reusable core. You can easily pass a config option to the app that simply avoids booting the router and you now have a widget with an API that will still properly invoke redraws.

Certainly, like in MVC you can opt to create a fat router/controller rather than a fat method-laden model, but to me the routing aspect is just decoration to a model's API methods and these methods should be invocable with the expectation that any necessary redraws (and optionally history routing) will happen without hand-wiring them together at every call site.

Having to hardcode the `path` of each filter into the view (and manually keep them in sync with your router config) is awkward and prone to breakage. Especially when you start getting into more parameterized/complex routes like `#/admin/users/:id/orderHistory?from=:from_date&to=:to_date`.

On one hand, the TodoMVC example is decoupled which is perhaps the least opinionated. On the other hand, it introduces additional boilerplate code because redraw() must be called everywhere and reduces DRY further by requiring you to duplicate the route paths in both the router and the app view. The question becomes, should the TodoMVC present a un unopinionated example filled with boilerplate or a more idiomatic/opinionated example that would be better suited as a demo of how to start more complex apps where the boilerplate would quickly become a nuisance and opinions/conventions must be introduced?


btw, with a `forEach` (as opposed to a `for`) your `r` is already closed over by the iterator callback, so you can drop the wrapping IIFE i think:

```js
var ret = {};

['all', 'active', 'completed'].forEach(function(r) {
	ret[r] = {
		path: r,
		onenter: function() {
			model.Todos.currentFilter = r;
			appView.redraw();
		}
	}
});

return ret;
```

**Comment by yosbelms** [comment]
I understand your routing approach, I'm aware of its benefits and how it promotes reusability. I just added the route module to the TodoMVC with little objectives in mind:

1. A minimal show of how to use the `route` component.
2. Few LOCs -as possible- without sacrificing readability.

I assume as TodoMVC is a small app and unlikely to grow in the future I can afford to write lean code whithout too much complication. Bigger architectures could be addressed further in docs, tutorials, articles, etc. Anyways, if the router is MVC-like or "decoration to a model API" makes no important differences to me, should be good to have another opinion. /summon @lawrence-dol.

> btw, with a forEach (as opposed to a for) your r is already closed over by the iterator callback,

Indeed. I think I was paranoiac when wrote that, lol

**Comment by leeoniya** [comment]
To be honest, the more I use the router's `init`, the less I want to use it. In fact it will probably be deprecated and removed from the docs. It is simply not necessary for it to live inside the router closure because the same thing can be accomplished externally:

```js
// shared context
var ctx = {
  app: null,
  view: null,
  router: null,
};

ctx.app = new App(state, ctx);
ctx.router = domvm.route(AppRouter, ctx);
ctx.view = domvm.view(AppView, ctx);

ctx.view.mount(document.body);
ctx.router.refresh();
```

Once `domvm.route()` is called, the router has initialized and can be used by the view for `href` gen or by the app for route inference. Having `init` within the route closure introduces confusion and offers no tangible benefits over the above pattern. It's really just unnecessary encapsulation that make the init order less obvious.

**Comment by yosbelms** [comment]
If no route is provided ,for example: `TodoMVC/index.html`, does `rouer.route().href` should be an empty string?  actually is equal to `/`.

**Comment by leeoniya** [comment]
there's definitely a need to decide on a spec/expectations.

in hash mode, it seems reasonable to return `#/`, `#`, or `/` since the client-side router is in a root path regardless of whether the served page url is `.html` or `.com`. What to display in the address bar is also unclear. There is a similar problem with how some browsers choose to display or not display a trailing slash on root domains but `window.location.href` will always return the trailing slash, as will most server-side routers when you access `/`. in history mode it's slightly different because of the distinction between what the server root is (prefix) and what the additional client-side route is.

what the router does currently may or may not be correct but these are the things to consider if we plan to spec it out in a way that's uniform and makes sense.

**Comment by yosbelms** [comment]
In that case it's fine to return `/` if the route is `''`, `#`, or `#/` in hash mode to be compliant with `window.location.href`, so, `/` means `empty route`. btw, actually `router.route().hash` in undefined, should it be?.

**Comment by leeoniya** [comment]
> btw, actually router.route().hash in undefined, should it be?

not sure if it should be undefined, but it can be null. hash mode is meant to be transparent, so everything after the hash is treated as it it was the full url. so a hash in hash mode would actually by the *second* hash.

`https://xyz.com/index.html#/abc/123?foo=bar#some-frag` would have a hash of `some-frag`, or maybe `#some-frag`, i forget. it's not a common thing though.

**Comment by yosbelms** [comment]
Hey @leeoniya, a `didEnter` hook in the router could be useful in many cases, Example:

```js
router.config({
    didEnter: function() {
        appView.redraw()
    }
})
```
any thought on this?

**Comment by leeoniya** [comment]
yes, `didEnter` and `didExit` will definitely be good additions and can complement the existing `willEnter` and `willExit`. wanna do a pull?

**Comment by leeoniya** [comment]
also to keep in mind, it might be a good idea to add the ability for some router hooks to return a thennable/Promise to defer additional callbacks. An immediately obvious case is where `onenter` makes an ajax request for state and `didEnter` invokes the redraw. Another case is where `willEnter` needs to re-auth a session token and defer `onenter` or prevent it and redirect to login, etc.

**Comment by yosbelms** [comment]
^ This feature deserves its own thread, IMO this put to much control flow in the router.

**Comment by leeoniya** [comment]
once you're using the route handlers to invoke redraw directly, you're already started down that path by delegating view sync away from your model/api. assuming your model/api can do asynchronous ajax/fetch, you should be able to rely on the router to sync ypur view. you could obviously chain redraws within `oneneter` but then you've lost the convenience of DRY didEnter redraw.

it' pretty trivial to handle thennable returns from the callbacks and let users decide where they want their control flow to take place.

**Comment by yosbelms** [comment]
See this example, with promises will look better, promises can do a good job on this. BTW I'm still not convinced about invoking async functions in the router. Should it be in the view, or in the router?

```
function didEnter() {
    appView.redraw()
}

router.config({ didEnter: didEnter })

return {
    active: {
        onenter: function() {
            fetch.get('todo/active').then(function(data){
                model.setData(data)
                didEnter()
            })
            // prevent didEnter execution
            return false
        }
    }
}
```

**Comment by leeoniya** [comment]
I don't think `didEnter`/`didExit` should be preventable, see https://github.com/leeoniya/domvm/pull/97#issuecomment-240570929

> Should it be in the view, or in the router?

It really depends on how you write apps. Since I write API-first, i would have a lot of this stuff live in a method-laden model since that would always be present regardless of routing or views. If you're designing with the expectation of a router and UI, then you could have it live in either the router or the view (since those constitute the core of your app if you have dumb json state), though my preference would be the router. Even this depends on whether/which views are persistent and how you cache/expire fetched state.

Bottom line is, domvm should remain flexible and opinionated where it's fast, cheap and architecturally possible. Handling async behavior in will/on callbacks is one of those things IMO.
